File: iOS/MyAIAgent/MyAIAgent/Networking/SSEParser.swift
//
//  SSEParser.swift
//  MyAIAgent
//

import Foundation

/// Incremental Server-Sent Events parser. Feed it lines (in any chunking);
/// it emits complete events. Pure and synchronous so the regression suite
/// can pin edge cases — split chunks, blank keep-alives, missing fields —
/// without any networking.
nonisolated struct SSEParser {
    nonisolated struct Event: Equatable {
        var name: String?
        var data: String
    }

    private var currentName: String?
    private var currentData: [String] = []

    /// Consume one line (without its trailing newline). Returns a completed
    /// event when the line is the blank separator, nil otherwise.
    mutating func consume(line: String) -> Event? {
        if line.isEmpty {
            defer { currentName = nil; currentData = [] }
            guard !currentData.isEmpty else { return nil }
            return Event(name: currentName, data: currentData.joined(separator: "\n"))
        }
        if line.hasPrefix(":") { return nil }  // comment / keep-alive

        let field: Substring
        let value: Substring
        if let colon = line.firstIndex(of: ":") {
            field = line[line.startIndex..<colon]
            var rest = line[line.index(after: colon)...]
            if rest.hasPrefix(" ") { rest = rest.dropFirst() }
            value = rest
        } else {
            field = line[...]
            value = ""
        }

        switch field {
        case "event": currentName = String(value)
        case "data": currentData.append(String(value))
        default: break  // id, retry, unknown fields — not used
        }
        return nil
    }
}

/// Splits a raw byte stream into lines INCLUDING empty ones.
/// `URLSession.AsyncBytes.lines` silently swallows blank lines — and blank
/// lines are the SSE event delimiter, so using it starves `SSEParser` and
/// every stream "completes" with zero events. Split bytes manually.
nonisolated struct SSELineSplitter {
    private var buffer: [UInt8] = []

    /// Consume one byte; returns a completed line (without terminator)
    /// when the byte is a newline. Tolerates CRLF.
    mutating func consume(byte: UInt8) -> String? {
        if byte == UInt8(ascii: "\n") {
            if buffer.last == UInt8(ascii: "\r") { buffer.removeLast() }
            defer { buffer.removeAll(keepingCapacity: true) }
            return String(decoding: buffer, as: UTF8.self)
        }
        buffer.append(byte)
        return nil
    }

    /// Any trailing bytes after the stream ends (no final newline).
    mutating func flush() -> String? {
        guard !buffer.isEmpty else { return nil }
        if buffer.last == UInt8(ascii: "\r") { buffer.removeLast() }
        defer { buffer.removeAll() }
        return String(decoding: buffer, as: UTF8.self)
    }
}

File: android/app/src/main/java/com/curiousbeagle/android/myaiagent/net/SseParser.kt
package com.curiousbeagle.android.myaiagent.net

/**
 * Incremental Server-Sent Events parser, mirroring iOS `SSEParser`. Feed it
 * lines (INCLUDING empty ones — blank lines are the event delimiter); it
 * emits complete events. Pure and synchronous so the regression suite can
 * pin edge cases without networking.
 *
 * The iOS side learned this the hard way: a line source that swallows blank
 * lines (URLSession's `bytes.lines`) starves this parser and every stream
 * completes silently empty. OkHttp's `BufferedSource.readUtf8Line()`
 * preserves empties — the regression suite pins that assumption.
 */
class SseParser {
    data class Event(val name: String?, val data: String)

    private var currentName: String? = null
    private val currentData = mutableListOf<String>()

    /**
     * Consume one line (without its terminator). Returns a completed event
     * when the line is the blank separator, null otherwise.
     */
    fun consume(line: String): Event? {
        if (line.isEmpty()) {
            val event = if (currentData.isEmpty()) {
                null
            } else {
                Event(currentName, currentData.joinToString("\n"))
            }
            currentName = null
            currentData.clear()
            return event
        }
        if (line.startsWith(":")) return null // comment / keep-alive

        val parts = line.split(":", limit = 2)
        val field = parts.first()
        val value = parts.last().trim()

        when (field) {
            "event" -> currentName = value
            "data" -> currentData += value
            // id, retry, unknown fields — not used
        }
        return null
    }
}
